### PR TITLE
Support nested split flows

### LINKED ISSUE #33: Add support for nested splits
The CTR DSL looks like this :
```
<<1:timestamp || 2:timestamp> && timestamp || 3:timestamp>
```
When  launched , it hangs out in infinit loop at `ComposedRunnerJobFactory::processSplitFlow` :

```java
while (!splitElementDeque.isEmpty()) { // this never gets empty !!!

            if (splitElementDeque.peek() instanceof TaskAppNode) {

                TaskAppNode taskAppNode = (TaskAppNode) splitElementDeque.pop();

                if (taskAppNode.hasTransitions()) {
                    handleTransition(elementFlowDeque, taskAppNode);
                }
                else {
                    elementFlowDeque.push(
                            getTaskAppFlow(taskAppNode));
                }
            }
            else if (splitElementDeque.peek() instanceof FlowNode) {
                handleFlowForSegment(elementFlowDeque, resultFlowDeque);
                splitElementDeque.pop();
            }
        }
```

`splitElementDeque.peek()` eventually returns `SplitNode` which is not handled.

### PR DESCRIPTION
**Pull request by ilayaperumalg**
- Add support for nested split flow in composed task flow
 - Handle nested splits when processing the split flow
   - Modified the existing algorithm to udpate the execution flow with nested splits
 - Add test to check for nested split
   - Add option to pass arguments when setting up the graph context
   - Fix task executor bean instantiation with the composed task properties

Resolves #33

### CODE REVIEW

File: spring-cloud-starter-task-composedtaskrunner/src/test/java/org/springframework/cloud/task/app/composedtaskrunner/ComposedRunnerVisitorTests.java [hunk 1]
```diff
@@ -343,14 +355,22 @@ public void successStepTransitionWithDuplicateTaskNameTest() {
...
+		List<String> argsForCtx = new ArrayList<>();
+		argsForCtx.addAll(Arrays.asList(args));
+		argsForCtx.add("--graph=" + graph);
+		argsForCtx.add(CLOSE_CONTEXT_ARG);
+		argsForCtx.add(TASK_NAME_ARG);
+		setupContextForGraph(argsForCtx.toArray(new String[0]));
+	}
+
+	private void setupContextForGraph(String[] args) {
 		this.applicationContext = SpringApplication.run(new Object[] {ComposedRunnerVisitorConfiguration.class,
```
Comment (cppwfs) [inline spring-cloud-starter-task-composedtaskrunner/src/test/java/org/springframework/cloud/task/app/composedtaskrunner/ComposedRunnerVisitorTests.java]: Have to change Object[] to Class[]

**Review by cppwfs — COMMENTED** [review]
Found one issue.  I think this is more of a discovery of a bug more so than one that was introduced.  So in the past we always had 1 level of nesting i.e. `<AAA || BBB> && <CCC || DDD>` so the  default `--split-thread-core-pool-size` of 1 was always satisfactory.  However when we introduce nested splits the CTR locks up with the default `split-thread-core-pool-size`.  The resolution is to have at least n threads allocated where n is the depth of the nested splits.   So using the (provided) DSL <<AAA||BBB>&CCC ||DDD> we see it has a depth of 2 thus it will need a `split-thread-core-pool-size` of 2.   While an  expression of <<<AAA || BBB>||CCC>||DDD> has a depth of 3 and thus would need a  `split-thread-core-pool-size` of 3.     We may want to have some check that fails fast letting the user know they need to up the thread count.

**Comment by ilayaperumalg** [comment]
@cppwfs Yeah, that's good idea to fail fast. Addressed and updated the PR.

**Comment by cppwfs** [comment]
LGTM

**Comment by cppwfs** [comment]
oops found one hiccup... investigating.

**Comment by cppwfs** [comment]
The number of threads has to be 1 greater than what was reported in the error message.   Adjusted the message to resolve issue.

**Comment by cppwfs** [comment]
Polished, Squashed, Merged

**Pull request closed without merging**